docs/explanations/document-storage.md

```diff
@@ -0,0 +1,154 @@
+# Bluesky Document Storage
+
+## Executive Summary
+
+For the first time in the ten-year history of the Bluesky project, the Bluesky
+core developers will soon recommend a change in how data and metadata from
+Bluesky documents should be stored.
+
+In the past, each Bluesky document has been stored as an individual record in a
+document database. The Bluesky document representation is well-suited to
+low-latency applications like live plotting, but it is not optimized for
+storage at rest. Common data analysis workloads, including batch reads or
+random access, are not efficient.
+
+Now, the new recommendation involves transforming the contents of the Bluesky
+documents for storage at rest in a form that is more efficient for data access.
+The implementation in development involves a change in technology from a
+document database (MongoDB) to SQL (PostgreSQL or SQLite).
+
+## Motivation for and problems with original storage
+
+From the earliest days in 2015, the Bluesky project has recommended storing
+Bluesky documents in MongoDB as the "canonical" datastore, and optionally
+exporting their content in other (sometimes lossy) formats. There have been
+multiple Python APIs to this underlying MongoDB storage, and more recently an
+HTTP API [^1]. But the underlying storage that these APIs write to and read
+from has been unchanged.
+
+A MongoDB database was created for each beamline instrument/endstation. Within
+each database, a MongoDB "collection" was created for each Bluesky document
+type (`run_start`, `run_stop`, `event`, ...). Documents emitted by the RunEngine
+are inserted in the collection matching their document type. The MongoDB
+document storage is being used like a logfile: it captures exactly what Bluesky
+emitted.
+
+There is an appealing simplicity to this design: it is straightforward to
+explain and implement. And it is optimal for "replaying" the stream of
+documents after an experiment---a use case that was emphasized in the early
+days. It is defensible as an initial implementation. However, it is very
+non-optimal for _batch reads_ and _random access_. These are critical
+shortcomings in a data store.
+
+In order to access a portion data from MongoDB as a table or an array, we
+effectively take a "transpose" of the Event documents to build a columnar
+representation of the data. The implementation is fairly complex, and thus
+expensive to debug and maintain. And the operation imposes a performance cost
+that becomes noticeable beyond ~100 events.
```

Comment:
A benchmark demonstrating the benefits of the SQL database vs. the MongoDB for data retrieval would be nice. I assume it's not too hard to compare both with identical dataset being written to both backends and fetched from either. External factor need to be considered, like comparing a local MongoDB against a remote/central PostgreSQL will be unfair due to latencies.